python/lsst/pipe/base/graphBuilder.py

```diff
@@ -817,7 +819,79 @@ def makeGraph(self, pipeline, collections, run, userQuery):
             classes.
         """
         scaffolding = _PipelineScaffolding(pipeline, registry=self.registry)
+
+        userQuery = self._addInstrumentToQuery(pipeline, userQuery)
+
         with scaffolding.connectDataIds(self.registry, collections, userQuery) as commonDataIds:
             scaffolding.resolveDatasetRefs(self.registry, collections, run, commonDataIds,
                                            skipExisting=self.skipExisting)
         return scaffolding.makeQuantumGraph()
+
+    def _addInstrumentToQuery(self, pipeline, query):
+        """Add an instrument restriction to the query if the pipeline has an
+        instrument and the query does not already have an instrument. If the
+        query does have an instrument verify that it matches the instrument in
+        the pipeline.
+
+        Parameters
+        ----------
+        pipeline : `Pipeline`
+            Same as the pipeline in `makeGraph`
+        query : `str`
+            The initial query string.
+
+        Returns
+        -------
+        query : `str`
+            The query string with an instrument restriction checked or added
+            if needed.
+
+        Raises
+        ------
+        RuntimeError
+            If the query contains an instrument restriction and the pipeline
+            has an instrument and the two do not match.
+        """
+        instrument = pipeline.getInstrument()
+        if instrument is not None:
+            if isinstance(instrument, str):
+                instrument = getInstrument(pipeline._pipelineIR.instrument, self.registry)
```

Comment:
isn't this `pipeline._pipelinesIR` exactly what `pipeline.getInstrument()` does above? I'm trying to work out why the argument here isn't `instrument`.

Reply:
I think it's fairly safe to add the instrument to the query without checking.  We also have another ticket open for checking that an instrument is present when it needs to be, which would be implemented down in `daf_butler` where we actually fully parse this expression and can do the check quite cleanly.  I hadn't though about checking for two different instruments there (since we won't _want_ to do that someday), but it might be worth doing now.